test/linalg/test_linalg_lstsq.fypp

```diff
@@ -0,0 +1,96 @@
+#:include "common.fypp"
+#:set RC_KINDS_TYPES = REAL_KINDS_TYPES + CMPLX_KINDS_TYPES
+! Test least squares solver
+module test_linalg_least_squares
+    use testdrive, only: error_type, check, new_unittest, unittest_type
+    use stdlib_linalg_constants
+    use stdlib_linalg, only: lstsq
+    use stdlib_linalg_state, only: linalg_state_type
+
+    implicit none (type,external)
+    private
+    
+    public :: test_least_squares 
+
+    contains
+    
+
+    !> Solve sample least squares problems
+    subroutine test_least_squares(tests)
+        !> Collection of tests
+        type(unittest_type), allocatable, intent(out) :: tests(:)
+        
+        allocate(tests(0))
+
+        #:for rk,rt,ri in REAL_KINDS_TYPES
+        #:if rk!="xdp"
+        tests = [tests,new_unittest("lease_squares_${ri}$",test_lstsq_one_${ri}$)]
+        #:endif
+        #:endfor
+
+    end subroutine test_least_squares
+
+    !> Simple polynomial fit
+    #:for rk,rt,ri in REAL_KINDS_TYPES
+    #:if rk!="xdp"
+    subroutine test_lstsq_one_${ri}$(error)
+        type(error_type), allocatable, intent(out) :: error
+
+        type(linalg_state_type) :: state
+        integer(ilp) :: rank
+
+        !> Example scattered data
+        ${rt}$, parameter :: x(*)  = real([1.0, 2.5, 3.5, 4.0, 5.0, 7.0, 8.5], ${rk}$)
+        ${rt}$, parameter :: y(*)  = real([0.3, 1.1, 1.5, 2.0, 3.2, 6.6, 8.6], ${rk}$)
+        ${rt}$, parameter :: ab(*) = real([0.20925829,  0.12013861], ${rk}$)
+
```

Comment:
An alternative, and maybe more general, test would be to:
1. Draw a random matrix `A` and a random vector `x`
2. Compute rhs as `b= matmul(A,x)
3. Perform `p=lstsq(A,b)`
4. Check if `p == x`

Reply:
[63d297e](https://github.com/fortran-lang/stdlib/pull/801/commits/63d297ecdc961c44f53b451f409828d69b9f8dda)